pkg/parca/parca.go

```diff
@@ -584,35 +584,44 @@ func runScraper(
 		},
 	)
 
-	parcaserver := server.NewServer(reg, version)
-	gr.Add(
-		func() error {
-			return parcaserver.ListenAndServe(
-				ctx,
-				logger,
-				flags.Port,
-				flags.CORSAllowedOrigins,
-				flags.PathPrefix,
-				server.RegisterableFunc(func(ctx context.Context, srv *grpc.Server, mux *runtime.ServeMux, endpoint string, opts []grpc.DialOption) error {
-					scrapepb.RegisterScrapeServiceServer(srv, m)
-					if err := scrapepb.RegisterScrapeServiceHandlerFromEndpoint(ctx, mux, endpoint, opts); err != nil {
-						return err
-					}
-					return nil
-				}),
-			)
-		},
-		func(_ error) {
-			ctx, cancel := context.WithTimeout(ctx, 30*time.Second) // TODO make this a graceful shutdown config setting
-			defer cancel()
-
-			level.Debug(logger).Log("msg", "server shutting down")
-			err := parcaserver.Shutdown(ctx)
-			if err != nil && !errors.Is(err, context.Canceled) {
-				level.Error(logger).Log("msg", "error shutting down server", "err", err)
-			}
-		},
-	)
+	{
+		parcaserver := server.NewServer(reg, version)
+		serveCtx, cancelServe := context.WithCancel(ctx)
+		gr.Add(
+			func() error {
+				return parcaserver.ListenAndServe(
+					serveCtx,
+					logger,
+					flags.Port,
+					flags.CORSAllowedOrigins,
+					flags.PathPrefix,
+					server.RegisterableFunc(func(ctx context.Context, srv *grpc.Server, mux *runtime.ServeMux, endpoint string, opts []grpc.DialOption) error {
+						scrapepb.RegisterScrapeServiceServer(srv, m)
+						if err := scrapepb.RegisterScrapeServiceHandlerFromEndpoint(ctx, mux, endpoint, opts); err != nil {
```

Comment:
The code within this function reads to me the way that we should also pass the `serveCtx` here.

Reply:
Not sure I understand. I think this `ctx` shadows the outer `ctx` and is a child of `serveCtx`

Reply:
The way I read the code within `RegisterScrapeServiceHandlerFromEndpoint` is that the passed context will be used to serve requests. Therefore the right context to use here is `serveCtx` as once cancelled with the `cancelServe` function those execution paths should cancel and not only when the main/parent context gets cancelled at the very end just before exiting the process.